### PR TITLE
Optimize sampling

### PR DESCRIPTION
**Pull request by grimoire**
sort logits of large vocab size would take long time on GPU.
This PR gather topk logits and slice scores after topp

llama2-13b (vocab_size=32000):

before

```
token throughput (completion token): 1462.612 token/s
token throughput (prompt + completion token): 3031.211 token/s
RPS (request per second): 6.131 req/s
RPM (request per minute): 367.831 req/min
```

after

```
token throughput (completion token): 1482.363 token/s
token throughput (prompt + completion token): 3072.144 token/s
RPS (request per second): 6.213 req/s
RPM (request per minute): 372.799 req/min
```

half the sampling time. Larger vocab_size would have better performance since radix sort has O(nk) complexity

### CODE REVIEW

**Review by RunningLeon — APPROVED** [review]
LGTM

File: lmdeploy/pytorch/engine/logits_process.py [hunk 1]
```diff
@@ -276,6 +276,11 @@ def __random_sampling(scores: torch.Tensor, indices: torch.LongTensor):
                 scores = _filter_topp_sorted(scores, top_p)
 
             softmax_scores = scores.softmax(1)
+
+            max_topk = sampling_inputs.max_top_k
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:280]: `max_top_k` can be 0, which means `top_k` sampling is skipped.

Comment (grimoire) [reply lmdeploy/pytorch/engine/logits_process.py:280]: top_k==0 is equal to top_k=vocab_size ?

Comment (lvhan028) [reply lmdeploy/pytorch/engine/logits_process.py:280]: Yes. Have we updated it to `vocab_size` when users' inputting `top_k` is 0?

Comment (grimoire) [reply lmdeploy/pytorch/engine/logits_process.py:280]: Never

**Review by lvhan028 — APPROVED** [review]
